### ISSUE TITLE
Quote auto closing is only triggered in strings

### ISSUE THREAD

**Issue by mleise** [Basile-z/dexed#402]
Maybe in response to https://github.com/BBasile/dexed/issues/29, the auto-closing of `'` only works inside of `"` and `` ` `` strings and the auto-closing of `"` only works in `'` strings. In particular it doesn't work outside of any open strings.
That reminds me: Some IDEs with auto-closing would overwrite (or skip rather) any existing closing quotes/braces you type when the caret is in front of them, so that you can either use the right arrow key to skip past auto-closed stuff or continue to type naturally. It's probably not as easy to implement as it sounds. It never is. Still very convenient feature.

**Comment by BBasile** [comment]
Yes i agree with everything. This feature is not as smart as it should. The [feature is based on the tokens](https://github.com/BBasile/dexed/blob/fee4a1b0d168f9d9af92f51dcd53fd27cc08490a/src/u_synmemo.pas#L2340) so there's room for improvement.

**Comment by BBasile** [comment]
I read again the first message of the report and maybe i've replied too fast. Actually for the pairs of chars auto closing works after entering the left one. So for example for single quote if you type `'` then `a`, you ends up with `'a'`.

**Comment by mleise** [comment]
> I read again the first message of the report and maybe i've replied too fast. Actually for the pairs of chars auto closing works after entering the left one. So for example for single quote if you type `'` then `a`, you ends up with `'a'`.

It seems to be more "diverse" than that. In an empty module just typing `'` would close the quote as expected. In an actual larger module it isn't closed at all (unless inside a `` ` `` or `"` string). And I don't know yet how to reproduce your case where it closes after one character typed. I have to say, it seems like a really cool feature to close characters automatically when they are completely defined. The kill-joy are octal definitions like `\1`, `\12` or `\123`: No way to know when the user wants to stop typing. Well, if you exempt this rare case it could still work out and Dexed would be the first and only IDE doing this. Sorry, got carried away.

**Comment by BBasile** [comment]
I remember also that a problem with this feature is that the editor texxt is lexed **after** the insertion of the left character. So the tokens following the cursor position can be wrong.

Otherwise i don't quite understand your last message @mleise. I think that to do something at this point it would be preferable to define clearly the patterns in pseudo language.

**Comment by BBasile** [comment]
I think that i didn't explain well how this work because you mention "when the user ends typing".

1. you type `'`.
2. a second `'` is automatically inserted just after.
3. the cursor position is set between the pair.
4. you type `a` (here you're not limited in length).
5. you end up with `'a'`.

**Comment by mleise** [comment]
> So for example for single quote if you type `'` then `a`, you ends up with `'a'`.

I see my mistake now. I read that as: Type `'a` and you end up with `'a'`.
In any case all of my +500 LOC files have this behavior where the closing `'` and `"` aren't inserted unless they are nested inside the the opposite thing. I.e. `'` is only closed within `"` and `"` is only closed in `'`. Weird. It is consistently reproducible in different bigger files for me (, like for example std.stdio).

**Comment by Basile-z** [comment]
I've managed to make the feature better. The new rules are

1. Never auto insert the closing char if the opening char is stuck to a token that's a literal, an idententifier.
2. Never insert the closing char if the cursor is located in a token of the type it's supposed to auto close, e.g in a string, never auto insert the closing double quote.
3. Never insert the closing chars within comments.
4. otherwise always insert the closing char.

The main problem was actually that the source was scanned after that the opening char was inserted.

**Comment by mleise** [comment]
Allow me to run some gedankenexperiments on this.
 1. I assume that ", an idententifier" part was supposed to be deleted? Then if "stuck to" means basically to the "right of", that would prevent **`foo!42(`** from auto-closing. Did you consider this?
 2. That handles the 2 cases of ``` ` ``` inside `"` and vice-versa, but not e.g. `"Please don'"`.
 3. Yes, but let me expand on this below ...

I'll try to back up my personal preference which goes like this:

 - `closing_char` is defined as the char used to close the token, that the cursor is in, i.e. ' inside character literals or " inside strings. It is undefined and unused in other contexts.
 - `is_escaped` is true, when the character to the left is `\`, but not after `\\`.
 - `column` is the cursor column starting at `0`.
 - Function calls are assumed to stop further processing and exit the routine.

```Pascal
if in comment and key is SINGLEQUOTE
    // Assume written word in comments. `'` often appears in English text alone,
    // while the other chars are usually paired.
    normalKeyPress()
if in (character_literal or normal_string) and is_escaped
    // Never auto-close any escaped characters
    normalKeyPress()
if in (character_literal or normal_string or raw_string)
    // Skip over closing char of ', " & ` when not in an escape sequence
    if right_char is closing_char and key is closing_char
        moveCursorRight()

// What do you do when the user presses `"` in the middle of a string?
// Special rules for splitting up strings in the middle to insert variables or for
// line-wrapping! (Or just handle as normalKeyPress() if you prefer.)
if in (normal_string or raw_string) and key is closing_char
    if column < soft_line_length_limit - 10
        // We are at a safe distance from the line length limit to insert a variable here
        // Inserts the closing char, " ~  ~ " and opening char (8 chars) and we know that at
        // least 2 follow = 10. Afterwards place the cursor between the two ~ where the user
        // can type the name of a variable to splice in.
        breakStringUp()
    else
        // When close to the end of the line, assume the user wants to wrap the line.
        // Inserts the closing char, space, '~', newline, indentation and opening char.
        // Places cursor in the next line in front of the opening char.
        lineWrapString()

// Otherwise insert char & closing char and move cursor between them as usual
autoCloseKey()

**Comment by Basile-z** [comment]
> I assume that ", an idententifier" part was supposed to be deleted?

No it was supposed to be

_1. Never auto insert the closing char if the opening char is stuck to a token that's a literal or an identifier._

**Comment by mleise** [comment]
Let me rephrase to see if I understood: When the cursor is at the beginning of a literal or an identifier, no auto-closing happens, based on the assumption that the closing char actually belongs somewhere *after* that literal/identifier. `int x = |2` ⟵ no auto-closing when typing `(` or `"`, etc.

**Comment by Basile-z** [comment]
Yes, you get `"2` and not `""2`. Have you tested the feature ? I believed that this feature was good now.

**Comment by mleise** [comment]
No, I didn't realize you were making releases all the while I was reporting bugs! I'll upgrade now and report back.

**Comment by Basile-z** [comment]
Because of  #415 I plan to release again today too.

**Comment by Basile-z** [comment]
Can you open a separate issue ?

**Comment by mleise** [comment]
Works as intended now! Since I'm still curious how my preferred rules would work out, I forked Dexed now. It wasn't as hard as the instructions made it look (manual patches to Lazarus and FPC are currently not required). Hopefully I can wrap my head around how the editor window works.
